File: text_cleanser.c
#define PCRE2_STATIC

#define PCRE2_CODE_UNIT_WIDTH 8

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <string.h>
#include "uthash/uthash.h"
#include "text_cleanser.h"
#include "preg_replace.h"

char *replace(char *re, char *replacement, char *subject) {

  char *output;
  output = preg_replace(re, replacement, subject);
  free(subject);
  return output;

}

char *cleanse_inbetween_commas(char *subject) {
  return replace("((?![a-zA-Z]+),(?=[a-zA-Z+]))", " ", subject);
}

char *cleanse_quotes(char *subject) {
  return replace("['\"`*]", "", subject);
}

char *cleanse_brackets(char *subject) {
  return replace("[\\(\\)\\[\\]\\{\\}]", " " , subject);
}

char *cleanse_html_entities(char *subject) {
  return replace("&#?[a-zA-Z0-9]+;", " ", subject);
}

char *cleanse_newline_characters(char *subject) {
  return replace("(\\n|\\x0D)", " ", subject);
}


char *cleanse_article_link_markup(char *subject) {
  return replace("\\[.+?\\]\\(\\d+\\s\\\".*?\\\"\\)", "", subject);
}

char *cleanse_non_ascii_characters(char *subject) {
  return replace("[^\\x00-\\x7F]", "", subject);
}

char *cleanse_preceding_special_characters(char *subject) {
  return replace("(^|\\s|\\.\\,)([\\@\\#\\$])+\\b", " ", subject);
}

char *cleanse_following_special_characters(char *subject) {
  return replace("\\b([\\!\\?\\%\\.\\,\\:\\;\\/\\$])+(\\Z|\\s|\\.|\\,)", " ",subject);
}

char *cleanse(char *subject) {
  char *input = strdup(subject);

  input = cleanse_non_ascii_characters(input);
  input = cleanse_article_link_markup(input);
  input = cleanse_quotes(input);
  input = cleanse_brackets(input);
  input = cleanse_inbetween_commas(input);
  input = cleanse_html_entities(input);
  input = cleanse_newline_characters(input);
  input = cleanse_preceding_special_characters(input);
  input = cleanse_following_special_characters(input);
  return input;
}

File: TextCleanser.php
<?php

class TextCleanser {

  public function cleanse($str) {

    $str = $this->cleanse_article_link_markup($str);
    $str = $this->cleanse_non_ascii_characters($str);
    $str = $this->cleanse_quotes($str);
    $str = $this->cleanse_brackets($str);
    $str = $this->cleanse_inbetween_commas($str);
    $str = $this->cleanse_html_entities($str);
    $str = $this->cleanse_newline_characters($str);
    $str = $this->cleanse_preceding_and_following_special_characters($str);

    return $str;
  }

  private function cleanse_inbetween_commas($str) {
    return preg_replace('/((?![a-zA-Z+]),(?=[a-zA-Z+]))/', ' ', $str);
  }

  private function cleanse_quotes($str) {
    return preg_replace('/[\'\"\`\*]/', '', $str);
  }

  private function cleanse_brackets($str) {
    return preg_replace('/[\(\)\[\]\{\}]/', ' ' , $str);
  }

  private function cleanse_html_entities($str) {
    return preg_replace("/&#?[a-zA-Z0-9]+;/", ' ',$str);
  }

  private function cleanse_newline_characters($str) {
    return str_ireplace(["\n", "\x0D"], " ", $str);
  }

  private function cleanse_article_link_markup($str) {
    return preg_replace('/\[.+?\]\(\d+\s\".*?\"\)/', "", $str);
  }

  private function cleanse_non_ascii_characters($str) {
    return preg_replace('/[^\x00-\x7F]/', '', $str);
  }

  private function cleanse_preceding_and_following_special_characters($str) {
    // chars to be removed only on beginning of word
    $patterns[] = '/(^|\s|\.\,)([\@\#\$])+\b/';

    // chars to be removed only on end of word
    $patterns[] = '/\b([\!\?\%\.\,\:\;\/\$])+(\Z|\s|\.|\,)/';

    return preg_replace($patterns, ' ', $str);
  }

}
